server/cluster/cluster.go

```diff
@@ -41,25 +40,24 @@ func NewServer(db db.ArgoDB, enf *rbac.Enforcer, cache *servercache.Cache, kubec
 	}
 }
 
-func (s *Server) getConnectionState(cluster appv1.Cluster, errorMessage string) appv1.ConnectionState {
+func (s *Server) getConnectionState(cluster appv1.Cluster, errorMessage string) (appv1.ConnectionState, string) {
 	if connectionState, err := s.cache.GetClusterConnectionState(cluster.Server); err == nil {
-		return connectionState
+		return connectionState.ConnectionState, connectionState.Version
 	}
 	now := v1.Now()
-	connectionState := appv1.ConnectionState{
-		Status:     appv1.ConnectionStatusSuccessful,
-		ModifiedAt: &now,
+	connectionState := servercache.ClusterConnectionState{
+		ConnectionState: appv1.ConnectionState{
+			Status:     appv1.ConnectionStatusSuccessful,
+			ModifiedAt: &now,
+		},
 	}
 
-	config := cluster.RESTConfig()
-	config.Timeout = time.Second
-	kubeClientset, err := kubernetes.NewForConfig(config)
-	if err == nil {
-		_, err = kubeClientset.Discovery().ServerVersion()
-	}
+	version, err := s.kubectl.GetServerVersion(cluster.RESTConfig())
```

Comment:
why this?

Reply:
I had to move it from https://github.com/argoproj/argo-cd/pull/2621/files#diff-02ca8f1595ce93b75591586acd227047L105 to avoid failing whole request if one cluster has incorrect credentials.

If we are able to get the cluster server version when it ensures that cluster is reachable and `kubeClientset.Discovery().ServerVersion()` no longer required.

Reply:
Thank explains why I don't understand .I don't think we should mix up connection state with server version, they are different things and should not be stored together like this.

Reply:
The caching was introduced because a single unreachable cluster makes `/cluster` API very slow. To avoid it we cache the most recent verification result. Also the verification is result of running `GetServerVersion`, so I think it is pretty much the same thing.